espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.java

```diff
@@ -0,0 +1,80 @@
+package io.appium.espressoserver.lib.handlers;
+
+import android.app.Activity;
+import android.support.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException;
+import io.appium.espressoserver.lib.helpers.ActivityHelper;
+import io.appium.espressoserver.lib.helpers.InvocationOperation;
+import io.appium.espressoserver.lib.model.MobileBackdoorMethod;
+import io.appium.espressoserver.lib.model.MobileBackdoorParams;
+
+import static io.appium.espressoserver.lib.helpers.AndroidLogger.logger;
+
+public class MobileBackdoor implements RequestHandler<MobileBackdoorParams, String> {
+
+    @Override
+    public String handle(final MobileBackdoorParams params) throws AppiumException {
+        logger.info("Invoking Backdoor");
+        Activity activity = ActivityHelper.getCurrentActivity();
+
+        List<InvocationOperation> ops = getBackdoorOperations(params);
+
+        if (ops.isEmpty()) {
+            throw new InvalidArgumentException("Please pass name(s) of methods to be invoked");
+        }
+
+        // First try to find the method in Application object
+        Object invocationResult = null;
+        try {
+            invocationResult = invokeBackdoorMethods(activity.getApplication(), ops);
```

Comment:
what if the result of invokeBackdoorMethods(activity.getApplication(), ops) is null? how do we distinguish this state from the state where no method has been found in the application?

Reply:
Currently, we do not.

Reply:
Then this should be probably fixed. Missing method is an exceptional situation, so it makes sense to handle an exception in such case rather than returning null

Reply:
I misread your comment previously. If a method is missing in the Application object, there will be an exception. Therefore invokeBackdoorMethods(activity.getApplication(), ops) will not return null.

Reply:
but the exception is anyway ignored. So in both cases invocationResult will be equal to null

Reply:
If there is a missing method in Application, we log this in . the the logcat to indicate the method was missing, and move on to find method in Activity.
```
11-28 21:53:05.479 10241 10271 I appium  : Invoking Backdoor
11-28 21:53:05.479 10241 10271 D appium  : Backdoor Method: messageView
11-28 21:53:05.479 10241 10271 D appium  : Invoking on: android.app.Application@4e7cde9
11-28 21:53:05.484 10241 10271 D appium  : Failed to find a suitable method using reflection
11-28 21:53:05.484 10241 10271 W System.err: io.appium.espressoserver.lib.handlers.exceptions.AppiumException: io.appium.espressoserver.lib.handlers.exceptions.AppiumException: {receiverClass=android.app.Application, error=No such method found: messageView(), receiverString=android.app.Application@4e7cde9, methodName=messageView}
```

Reply:
Understanding here is: A user will define backdoor in either Application or in Activity. We give priority to Application. If its not found in Application, we log it and move on to Activity, if its not found there, we throw exception.

Reply:
Possibly, we can store exception if method not found in Application. Later, if the method is not found in Activity as well, then we can inform the client in a better way by combining both messages?

Reply:
> execution of this method was successful and returned null
results will return "<VOID>"  in this case. 
`invocationResult = op.apply(invokeOn);` will either return a result or will throw exception

Reply:
> results will return <VOID> in this case.

It will only return `<VOID>` if the result of the method is declared as void. Otherwise null can be easily returned.

> will either return a result or will throw exception

do  we distinguish exceptional cases if the method was not found and if it was found but there was an exception on its execution?